File: C/Linked List/Doubly Circular Linked LIst/DoublyCL05.c
/*
=====================================================================

Name            : Doubly Circular Linked List - Display Function
Program Number  : 005

Description     : This program demonstrates traversal and
                  display of a Doubly Circular Linked List
                  using a do-while loop.

                  InsertFirst() and InsertLast() are used
                  to create the linked list and Display()
                  traverses all nodes exactly once.

Concepts Used   : Doubly Circular Linked List,
                  Display Function,
                  Traversal,
                  do-while Loop,
                  Dynamic Memory Allocation

Input           : Node data

Output          : Displays all nodes of the linked list

Date            : July 2026
Language        : C
Author          : Atharv Tushar Bhosale

=====================================================================
*/

// Doubly Circular Linked list

#include<stdio.h>
#include<stdlib.h>

#pragma pack(1)
struct node
{
    int data;
    struct node *next;
    struct node *prev;
};

typedef struct node NODE;
typedef struct node* PNODE;
typedef struct node** PPNODE;

void Display(PNODE first, PNODE last)
{   
    if(first == NULL && last == NULL)
    {
        return;
    }


    printf("<=>");
    do
    {
        printf("| %d | <=> ",first -> data);
        first = first -> next;
    }while(first != last -> next);
    printf("\n");
}

int Count(PNODE first, PNODE last)
{
    return 0;
}

void InsertFirst(PPNODE first, PPNODE last, int iNo)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    newn -> data = iNo;
    newn -> next = NULL;
    newn -> prev = NULL;

    if(*first == NULL && *last == NULL)
    {
        *first = newn;
        *last = newn;
    }
    else
    {
        newn -> next = *first;
        (*first )-> prev = newn;
        *first = newn;
    }
    (*last) -> next = *first;
    (*first) -> prev = *last;
}

void InsertLast(PPNODE first, PPNODE last,int iNo)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    newn -> data = iNo;
    newn -> next = NULL;
    newn -> prev = NULL;

    if(*first == NULL && *last == NULL)
    {
        *first = newn;
        *last = newn;
    }
    else
    {
        (*last) -> next = newn;
        newn -> prev = *last;
        *last = newn;
    }
    (*last) -> next = *first;
    (*first) -> prev = *last;
}

void InsertAtPos(PPNODE first, PPNODE last, int iNo, int iPos)
{

}

void DeleteFirst(PPNODE first, PPNODE last)
{

}

void DeleteLast(PPNODE first, PPNODE last)
{

}

void DeleteAtPos(PPNODE first, PPNODE last, int iPos)
{

}

int main()
{
    PNODE head = NULL;
    PNODE tail = NULL;

    InsertFirst(&head,&tail,51);
    InsertFirst(&head,&tail,21);
    InsertFirst(&head,&tail,11);

    InsertLast(&head,&tail,101);
    InsertLast(&head,&tail,111);
    InsertLast(&head,&tail,121);

    Display(head,tail);

    return 0;
}

/*
=====================================================================

Sample Output :

<=> | 11 | <=> | 21 | <=> | 51 | <=> | 101 | <=> | 111 | <=> | 121 | <=>

=====================================================================

Functions Implemented :

1. Display()
2. InsertFirst()
3. InsertLast()

Pending Functions :

1. Count()
2. InsertAtPos()
3. DeleteFirst()
4. DeleteLast()
5. DeleteAtPos()

=====================================================================

Notes :

1. Display uses a do-while loop.
2. Traversal starts from first node.
3. Traversal stops when first node is reached again.
4. Empty linked list is handled separately.
5. Circular structure is preserved throughout traversal.

=====================================================================
End of Program
=====================================================================
*/

File: Java/Linked List/Doubly Circular Linked List/DoublyCL05.java
/*
===============================================================================
Program Name    : Doubly Circular Linked List - Count Function
Program Number  : 005
File Name       : DoublyCL05.java

Description     : This program demonstrates the Count()
                  operation in a Doubly Circular Linked List
                  using Java. The function traverses the 
                  linked list and returns the total number
                  of nodes present in the list.

Date            : July 2026
Language        : Java
Author          : Atharv Tushar Bhosale
===============================================================================

Concepts Used:
1. Class and Object
2. Self-Referential Class
3. Doubly Circular Linked List
4. Dynamic Memory Allocation
5. do-while Loop
6. Node Counting

Key Learning Points:
- Traverse a Doubly Circular Linked List.
- Count total nodes in the list.
- Use do-while loop for circular traversal.
- Return the count of nodes.

===============================================================================
*/

class node
{
    public int data;
    public node next;
    public node prev;

    node(int no)
    {
        this.data = no;
        this.next = null;
        this.prev = null;
    }
}

class DoublyCL
{
    private node first;
    private node last;
    private int iCount;

    public DoublyCL()
    {
        first = null;
        last = null;
        iCount = 0;
    }

    public void Display()
    {
        if(first == null && last == null)
        {
            return;
        }

        node temp = first;

        do
        {
            System.out.print("| " + temp.data + " | <=> ");
            temp = temp.next;
        }while(temp != last.next);

        System.out.println();
    }

    public int Count()
    {
          return iCount;
    }
  
    public void InsertFirst(int iNo)
    {
        node newn = new node(iNo);

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else
        {
            newn.next = first;
            first.prev = newn;
            first = newn;
        }

        last.next = first;
        first.prev = last;

        iCount++;
    }

    public void InsertLast(int iNo)
    {
        node newn = new node(iNo);

        if(first == null && last == null)
        {
            first = newn;
            last = newn;
        }
        else
        {
            last.next = newn;
            newn.prev = last;
            last = newn;
        }

        last.next = first;
        first.prev = last;

        iCount++;
    }
}

class DoublyCL04
{
    public static void main(String A[])
    {
        DoublyCL dobj = new DoublyCL();
        int iRet = 0;
      
        dobj.InsertFirst(51);
        dobj.InsertFirst(21);
        dobj.InsertFirst(11);

        dobj.InsertLast(101);
        dobj.InsertLast(111);

        dobj.Display();

        iRet = dobj.Count();
        System.out.println("Number of Nodes: "+iRet);
    }
}

/*

EXPECTED OUTPUT:-

| 11 | <=> | 21 | <=> | 51 | <=> | 101 | <=> | 111 | <=>

Number of Nodes : 5
*/
